File: extract-color-palettes.py
#!/usr/bin/env python3
import json
import os
import re
import base64
from pathlib import Path
from PIL import Image
from io import BytesIO
from collections import Counter

def extract_png_from_svg(svg_path):
    """Extract base64 PNG data from SVG file."""
    with open(svg_path, 'r', encoding='utf-8') as f:
        svg_content = f.read()

    # Find base64 PNG data in href attribute
    match = re.search(r'href="data:image/png;base64,([^"]+)"', svg_content)
    if not match:
        raise ValueError(f"No PNG data found in {svg_path}")

    png_base64 = match.group(1)
    png_bytes = base64.b64decode(png_base64)
    return png_bytes

def extract_color_palette(image, max_colors=5, min_pixels=100):
    """Extract dominant colors from image."""
    # Convert to RGBA if needed
    if image.mode != 'RGBA':
        image = image.convert('RGBA')

    # Get pixel data
    pixels = list(image.getdata())

    # Filter out transparent pixels
    opaque_pixels = [
        (r, g, b, a) for r, g, b, a in pixels
        if a > 127  # Only count pixels with >50% opacity
    ]

    if not opaque_pixels:
        return []

    # Quantize colors to reduce noise (group similar colors)
    quantized = []
    for r, g, b, a in opaque_pixels:
        # Round to nearest 15 to group similar colors (wider range)
        qr = round(r / 15) * 15
        qg = round(g / 15) * 15
        qb = round(b / 15) * 15
        quantized.append((qr, qg, qb))

    # Count color frequencies
    color_counts = Counter(quantized)

    # Filter by minimum pixel threshold and get top colors
    filtered = [
        (color, count) for color, count in color_counts.items()
        if count >= min_pixels
    ]
    filtered.sort(key=lambda x: x[1], reverse=True)

    # Format as hex colors with metadata
    palette = []
    for (r, g, b), count in filtered[:max_colors]:
        hex_color = f'#{r:02X}{g:02X}{b:02X}'
        percentage = (count / len(opaque_pixels)) * 100
        palette.append({
            'color': hex_color,
            'pixelCount': count,
            'percentage': round(percentage, 2)
        })

    return palette

def main():
    icons_dir = Path(__file__).parent / 'public' / 'assets' / 'resco-icons'

    # Find all SVG files
    svg_files = sorted(icons_dir.glob('*.svg'))
    print(f"Found {len(svg_files)} SVG files. Processing...\n")

    output = {}

    for svg_path in svg_files:
        resco_name = svg_path.stem  # filename without .svg

        try:
            print(f"Processing: {resco_name}")

            # Extract PNG and open as image
            png_bytes = extract_png_from_svg(svg_path)
            image = Image.open(BytesIO(png_bytes))

            # Extract color palette
            palette = extract_color_palette(image, max_colors=5, min_pixels=100)

            output[resco_name] = {
                'colors': palette,
                'colorCount': len(palette)
            }

            print(f"  ✓ Found {len(palette)} colors\n")

        except Exception as e:
            print(f"  ✗ Error: {e}\n")

    # Write output to JSON file
    output_path = icons_dir / 'color-palettes.json'
    with open(output_path, 'w') as f:
        json.dump(output, f, indent=2)

    print(f"\n✓ Color palettes saved to: {output_path}")
    print(f"\nSummary:")
    print(f"- Total rescos processed: {len(output)}")
    print(f"- Output file: color-palettes.json")

    # Print a sample
    if output:
        first_resco = list(output.keys())[0]
        print(f"\nExample ({first_resco}):")
        for color_info in output[first_resco]['colors']:
            print(f"  {color_info['color']} - {color_info['percentage']}%")

if __name__ == '__main__':
    main()

File: extract-color-palettes.js
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

// Utility function to decode base64 and extract PNG data
function extractPNGFromSVG(svgPath) {
  const svgContent = fs.readFileSync(svgPath, 'utf8');
  const base64Match = svgContent.match(/href="data:image\/png;base64,([^"]+)"/);

  if (!base64Match) {
    throw new Error(`No PNG data found in ${svgPath}`);
  }

  return Buffer.from(base64Match[1], 'base64');
}

// Simple PNG parsing to get pixel data
function parsePNG(buffer) {
  let offset = 8; // Skip PNG signature
  const width = buffer.readUInt32BE(offset + 8);
  const height = buffer.readUInt32BE(offset + 12);

  let pixelData = [];
  let chunkOffset = 16; // Skip PNG header chunk

  while (chunkOffset < buffer.length - 12) {
    const chunkLength = buffer.readUInt32BE(chunkOffset);
    const chunkType = buffer.toString('ascii', chunkOffset + 4, chunkOffset + 8);

    if (chunkType === 'IDAT') {
      const compressedData = buffer.slice(chunkOffset + 8, chunkOffset + 8 + chunkLength);
      const imageData = zlib.inflateSync(compressedData);

      // Parse scanlines (simple RGBA parsing)
      let pixels = [];
      let idx = 0;
      for (let y = 0; y < height; y++) {
        idx++; // Skip filter byte
        for (let x = 0; x < width; x++) {
          const r = imageData[idx++];
          const g = imageData[idx++];
          const b = imageData[idx++];
          const a = imageData[idx++] || 255;

          // Only include non-transparent pixels
          if (a > 127) {
            pixels.push({ r, g, b, a });
          }
        }
      }

      return pixels;
    }

    chunkOffset += chunkLength + 12;
  }

  return [];
}

// Extract dominant colors from pixel data
function extractColorPalette(pixels, maxColors = 5, minPixels = 100) {
  if (pixels.length === 0) return [];

  // Count color frequencies
  const colorMap = new Map();

  for (const pixel of pixels) {
    // Round colors to nearest 5 to group similar colors
    const key = `${Math.round(pixel.r / 5) * 5},${Math.round(pixel.g / 5) * 5},${Math.round(pixel.b / 5) * 5}`;
    colorMap.set(key, (colorMap.get(key) || 0) + 1);
  }

  // Filter by minimum pixel threshold and sort by frequency
  const palette = Array.from(colorMap.entries())
    .filter(([_, count]) => count >= minPixels)
    .sort((a, b) => b[1] - a[1])
    .slice(0, maxColors)
    .map(([color, count]) => {
      const [r, g, b] = color.split(',').map(Number);
      return {
        color: `#${((1 << 24) + (r << 16) + (g << 8) + b).toString(16).slice(1).toUpperCase()}`,
        pixelCount: count,
        percentage: ((count / pixels.length) * 100).toFixed(2)
      };
    });

  return palette;
}

// Main script
const iconsDir = path.join(__dirname, 'public/assets/resco-icons');
const output = {};

const files = fs.readdirSync(iconsDir)
  .filter(f => f.endsWith('.svg'))
  .sort();

console.log(`Found ${files.length} SVG files. Processing...\n`);

for (const file of files) {
  const filePath = path.join(iconsDir, file);
  const rescoName = path.basename(file, '.svg');

  try {
    console.log(`Processing: ${rescoName}`);
    const pngBuffer = extractPNGFromSVG(filePath);
    const pixels = parsePNG(pngBuffer);
    const palette = extractColorPalette(pixels, 5, 100);

    output[rescoName] = {
      colors: palette,
      totalPixels: pixels.length
    };

    console.log(`  ✓ Found ${palette.length} colors\n`);
  } catch (error) {
    console.error(`  ✗ Error: ${error.message}\n`);
  }
}

// Write output to JSON file
const outputPath = path.join(iconsDir, 'color-palettes.json');
fs.writeFileSync(outputPath, JSON.stringify(output, null, 2));

console.log(`\n✓ Color palettes saved to: ${outputPath}`);
console.log(`\nSummary:`);
console.log(`- Total rescos processed: ${Object.keys(output).length}`);
console.log(`- Output file: color-palettes.json`);
